### PR TITLE
sys-process/runit: apply gcc-14/clang-18 ftbfs patch from debian

### PR DESCRIPTION
**Pull request by clan**
1. add patch from debian to fix compile issuing
2. update communication-files as modified in debian patch 0007-move-communication-files.patch

Closes: https://bugs.gentoo.org/938262
 
---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

**Comment by powerman** [comment]
@clan Can you please explain rationale why you decided to go same way as -r3 (add hard dependency on openrc, install new `/etc/runit/{1,2,3,…}` files, move some files to /run, …) instead of keeping runit same as it was many years up to -r1? Why don't take patches provided in runit's maillist, which does not include all these changes and just fix issues related to newer GCC?

**Comment by clan** [comment]
@powerman 
1. I don't have opinion for dependency on openrc, i.e., it can be removed
2. 2.1.2-r1 have /etc/runit/{1,2,3,…} installed too.
3. I aggress what said in debian patch 0007-move-communication-files.patch, /run is always writable while /etc is not
4. I've checked all the patch from debian, 0007 is the only patch changed some behaviour, the other patches either fix some compiler warnings or add some enhanced feature.

The idea to use debian's patchset is to keep things simple and patch in single place. If new non gentoo specific bug found later, I'd like to contribute the patch to debian too. In fact, it is much better if runit can accept these patches, but there is no official github available.

**Comment by powerman** [comment]
To me this doesn't sounds like a good idea. I understood your points and I agree they're valid ones. But the problem is this way we won't have "the runit" anymore.

It's clear enough some developers are really outstanding ones - and DJB is one of them. This is because other developer are unable to write good code **this way**. Gerrit Pape is one of few peoples who actually can write good code this (DJB-like) way. But unlikely all authors of these 27 patches are like DJB or Pape.

Sometimes it's unavoidable to patch their software - GCC issues is a good reason, adding features are most likely bad reason - but sometimes we just need that feature and have no other option. Here and now I don't think this is the case: people who needs more features in runit have an option to use s6 instead, and I'm not sure about Debian users but it's clear Gentoo users won't need extra features in runit - because all were happy enough with -r1 version of ebuild without any extra patches.

I can't speak for everyone, but for me this (adding 27 patches of unknown quality most of which no one of Gentoo users needs for "ease of maintenance" reason) is just not acceptable. If this will be merged then I'll just give up on using runit from portage tree and switch to my overlay without any patches.

**Comment by eli-schwartz** [comment]
We had to delete the -r3 ebuild precisely because it was such a disaster pile on account of trying to do many things at once, some of which violated the expectations of users of the package.

I am thoroughly against turning around and reinstating the exact problem we just dropped from the tree. No more functionality changes. The commit message for this PR says:

> apply gcc-14/clang-18 ftbfs patch from debian

and applying patches other than an ftbfs patch means that the commit message is UNTRUTHFUL.

**Comment by eli-schwartz** [comment]
Anyone is free to open a new PR that ONLY applies compiler error fixes and doesn't blindly apply "all patches which debian decides to apply".

**Comment by clan** [comment]
ok, it looks like:
1. we need to add a USE flag such as vanilla (or other better flag)
2. if we don't need the extra (non compiler related) patches from debian, then we need to maintain patches in gentoo seperately,

I can accept either way, how and where to decide this?

**Comment by clan** [comment]
> Anyone is free to open a new PR that ONLY applies compiler error fixes and doesn't blindly apply "all patches which debian decides to apply".

Just to be clear, I don't agree the words blindly. I do review all the patches, and I think it is not unacceptable.
I think we need to find a place to discuess which patch is not needed in gentoo.
Again, since there are no active maintainer right now, where to discuss and determine the next direction of runit in Gentoo?

**Comment by powerman** [comment]
> 1. we need to add a USE flag such as vanilla (or other better flag)

Yes, this is one of possible ways how to handle this. Maybe add vanilla flag. Maybe instead add feature-related flags. Personally I think in any case default should be no extra features (either enabled vanilla or disabled feature-flags), but this isn't very critical.

> 2. if we don't need the extra (non compiler related) patches from debian, then we need to maintain patches in gentoo seperately,

I'm not sure what do you mean as "maintain". Copying about 4 needed patches into `files/` subdir? Sure. Do some unknown amount of extra work which differs a lot from needs to monitor for changes in ubuntu patchset, review changes in that patchset, somehow manage unacceptable changes, and update ebuild - then no, I don't think there is noticeable amount of extra work if we'll not use the whole patchset and instead take only patches actually needed in Gentoo. If some required patches will depends on not required patches - then it'll take some work to cleanly re-apply them without not required patches. Probably it worth looking into 4 patches provided in runit maillist yesterday, chances are they are not depend on any others and just do the work - fix GCC issues and nothing else. (BTW, to me runit maillist looks like much more trusted source for such a patches than Ubuntu patchset.)

Actually, to me "extra work" is an attempt to maintain all these not needed changes from rest 20+ patches.

> I can accept either way, how and where to decide this?

I'm not a Gentoo dev, I'm just a proxy-maintainer for a few packages just like you, so I may be wrong about this, but: at this point it's up to you. You're now trying to maintain it - so you're decision maker now. At least as long as your decisions are in line with general Gentoo way. The final discussion will anyway happens in PR (just like now) and final decision is anyway by some Gentoo dev who will either accept or decline the PR.

> I don't agree the word blindly. I do review all the patches, and I think it is not unacceptable.

Well, word "blindly" might sounds too harsh, but adding 27 patches instead of 4 without (expected in this case) extra work is also a bit shocking, to be honest. And "extra work" here usually means:
- Consider is each patch actually needed/wanted/useful for Gentoo (not Debian/Ubuntu/etc.) users.
- Consider how changes of each patch will affect existing users. If it may break something then provide an upgrade path (some helpful messages in ebuild explaining what/how needs to be changed manually by user or what happens if user won't do anything).
- Consider which changes must be applied unconditionally and which must be enabled by some USE flag (and it that flag should be enabled by default).
- Provide separate PRs for adding each such patch-feature to make it possible to easily review/discuss it before merging.

This is fair amount of work, and as long as no Gentoo users actually asked for all these changes - it looks like an useless work.

**Comment by clan** [comment]
@powerman thanks
l'll review the debian patchset and patch in maillist one by one, and try to merge the compiling issue into gentoo first.

for other issue such as patch 0007 and patch 0013 (SIGPWR support), let's discuss later.

i'll setup a repository on github for this purpose.

**Comment by powerman** [comment]
@clan Please ping me in your future runit-related PRs and/or contact outside of github if you want, I'll be glad to help/discuss this. I'm using runit for 20.5 years and love it.

**Comment by clan** [comment]
@powerman I've setup a github repo for runit at: https://github.com/clan/runit/tree/ftbfs
ebuild patch is available at https://github.com/clan/gentoo/tree/runit

please help test whenever you have time. We can create a issue on my runit repo and discuss there if any questions or suggestions , thanks.

**Comment by thesamesam** [comment]
Hi both, I've quickly read over the above discussion and I think I agree with what Eli (and the rest of you) have said.

TL;DR:
* I like the idea of a proper fork repo where the Debian/Ubuntu patches are cherry-picked (with attribution and metadata, of course) where they're useful, with rationale for including them
* It is OK with me if you do a quick PR to get things working again and tidy up, adopt the package (as I trust you both will stick around), and then do followups with more develpoment.
* Please do make a new PR when you're ready for the big changes too.

**Comment by clan** [comment]
@thesamesam thanks for the comment.
1. I've complete the compile error & warning fixed (in order) in my repo
2. it's hard to apply patch from debian directly (which is bug based), while my patch are compile error & warning based. Right now, I've do diff check w/ source with debian patches applied to make sure nothing important are missed.
3. since there are no official github repo for runit, the patches will be bigger which can't fit in FILESDIR, please give some advice on howto distribute the patches properly.

P.S. unrelated to this runit issue, I also submit two PR on app-crypt/cfssl & www-server/nginx, both w/ small modification, what else can I do to let them to be merged (or do more modification if have)?

**Comment by eli-schwartz** [comment]
Some people host patches by creating a github repo containing an imported upstream tarball, with additional patches applied on top, and then all the patches exported with git format-patch, tarred up, and uploaded as a Github Releases asset.

**Comment by clan** [comment]
ok, i see. Currently I post all the formated patches in FILESDIR for develop and test, once it's confirmed and ready to release, I'll upload to github for final release. thanks.

**Comment by clan** [comment]
@eli-schwartz @thesamesam 

The patches for runit are almost ready, see https://github.com/clan/runit/compare/master...ftbfs.v3. There are totally  20 patches for now. Currently the patch are divided into the following categories (by order of patches):
1. helpful for development & debug, e.g, provide full command line and result, to check what's going on
2. required, build will failed if not included
3. patch from author w/ new feature provided (don't change existing behaviour)
4. documentation change: w/ tiny (unlikely to cause problem) modification of source code
5. source code changed to silience compiler warnings
6. source code changed to handle obsolete C function

The fact is that there is no active upstream for runit for now, so what's the principle to determine which patch to be included or not. I'd like to include all the patches, if it's not ok, 1/2 must be included, it's best to include 3/4, the rest are up to your decision? Thanks.

BTW Is it possible to maintain runit under Gentoo? Of course, I don't know the author's opinion, just ask for a possibility.

**Comment by thesamesam** [comment]
Sorry, this was on my list to reply to today (had the tab open still).

I suggest we do the following:
1. You create a repo (as you have done) for runit;
2. You have a branch with the original upstream sources
3. You have a branch with your patches and you cut a new tag with the patches that are required to _make runit build_ and don't change any "user-observable functionality"
4. We land a PR in gentoo which pulls in those patches, including you adopting/maintaining runit in Gentoo, and drop last-rites
5. You then work with @powerman to discuss the remainder of the patches as to whether they should be included or not, and make further PRs to ::gentoo based on that

**Comment by clan** [comment]
> Sorry, this was on my list to reply to today (had the tab open still).
> 
> I suggest we do the following:
> 
> 1. You create a repo (as you have done) for runit;
> 2. You have a branch with the original upstream sources
> 3. You have a branch with your patches and you cut a new tag with the patches that are required to _make runit build_ and don't change any "user-observable functionality"
> 4. We land a PR in gentoo which pulls in those patches, including you adopting/maintaining runit in Gentoo, and drop last-rites
> 5. You then work with @powerman to discuss the remainder of the patches (new features, maybe doc changes) as to whether they should be included or not, and make further PRs to ::gentoo based on that

cool, thanks.

**Comment by powerman** [comment]
> 5. You then work with @powerman to discuss the remainder of the patches (new features, maybe doc changes) as to whether they should be included or not, and make further PRs to ::gentoo based on that

This was already done. The result of my review for all these patches is available starting from this comment: https://github.com/clan/runit/issues/2#issuecomment-2324365861. Overall resume: all patches looks good to me. But I'm not sure is it good idea to merge this one:

> 3. patch from author w/ new feature provided (don't change existing behaviour)

It is really made by runit's author and it is clean, but the feature it adds looks a bit weird to me.
It adds a new feature to work around decade old Debian issue with slow vserver shutdown (with unknown root cause) **by disabling `sync` on halt/reboot**. It may be still actual today, say, for docker containers, but as we have no bugreports from Gentoo users we can't be sure and have no possibility to research actual root cause. So (mis)using this feature on a real server (i.e. not in a docker container etc.) may lead to data loss on halt/reboot. So, is it good (works around some real issue still actual today) or bad (provides a new way for shooting yourself in the foot) is at least questionable.

**Comment by clan** [comment]
> > 3. patch from author w/ new feature provided (don't change existing behaviour)
> 
> It is really made by runit's author and it is clean, but the feature it adds looks a bit weird to me. It adds a new feature to work around decade old Debian issue with slow vserver shutdown (with unknown root cause) **by disabling `sync` on halt/reboot**. It may be still actual today, say, for docker containers, but as we have no bugreports from Gentoo users we can't be sure and have no possibility to research actual root cause. So (mis)using this feature on a real server (i.e. not in a docker container etc.) may lead to data loss on halt/reboot. So, is it good (works around some real issue still actual today) or bad (provides a new way for shooting yourself in the foot) is at least questionable.

I've some different opinion here, but it's OK to go w/o without this patch for now or forever (until someone ask).

Technique discuss: this patch just add an option which let user (or sysadmin) can do faster restart etc. if they want to, we can't prevent user doing this in real machine (which will cause the problem you said). It's off by default so won't cause any new problem if user just do a regular upgrade. User must know what they are doing if they choose to turn on this feature. Such as we can't prevent user do 'echo b > /proc/sysrq-trigger' if they are root user.
